rmw_connext_cpp/src/functions.cpp

```diff
@@ -164,62 +212,73 @@ rmw_create_publisher(
     rmw_set_error_string("callbacks handle is null");
     return NULL;
   }
-  std::string type_name = std::string(callbacks->package_name) + "::msg::dds_::" +
-    callbacks->message_name + "_";
+  std::string type_name = _create_type_name(callbacks, "msg");
+  // Past this point, a failure results in unrolling code in the goto fail block.
```

Comment:
I looked around for the "best" practice for this pattern and I saw this recommendation as an alternative to `goto fail`:

```
do {
  a = ...;
  if (!a) break;
  b = ...;
  if (!b) break;
  // and so on
  return success;
} while (false);
// clean up
return failure;
```

Which seems like what you're suggesting, but it seemed to be that people generally recommended the goto fail style. The example everyone kept pointing back to was the Linux kernel.

I'm open to changing it, but I'd like to know what others think. @codebot specifically, you've done more of this style of coding, what do you think?